### PR TITLE
Feature: add/edit Hero profile data

### PR DESCRIPTION
**Pull request by rpollock03**
Skeleton component and functionality for a hero to add or update their profile data.

Logged in users view a basic component that looks like this:

Screenshot 2022-09-27 at 17 29 21

Clicking the edit button displays a form allowing user to add, or update their profile information:

Screenshot 2022-09-27 at 17 29 48

Note: the fields of the form are pre-populated with any existing user profile data found on firebase (the document ID on firebase is the same as the userId). In this way, a new user can add their profile information, or an existing user can update profile information they have previously sent.

Includes an add experience button which allows the user to add as many experiences as needed to the experience array.

Screenshot 2022-09-27 at 16 46 25

Firebase document consistent with the Hero data type:

Screenshot 2022-09-27 at 17 44 49

### CODE REVIEW

**Comment by OliwierOST** [comment]
Screen Shot 2022-10-02 at 12 06 33 AM

Above is redundant, information about logged in user can be found here

Screen Shot 2022-10-02 at 12 06 55 AM

File: components/EditProfile.tsx [hunk 1]
```diff
@@ -0,0 +1,246 @@
+import { Formik, Form, FieldArray } from "formik"
+import { useFirestore, useUser } from "reactfire"
+import { doc, setDoc } from "firebase/firestore"
+import { Caption, FormField } from "./Form"
+
+import { Hero } from "../storage/hero"
+
+export default function EditProfile({ hero }): JSX.Element {
```
Comment (OliwierOST) [inline components/EditProfile.tsx:8]: The type of props is not specified for `EditProfile`

Should declare an `interface Props {}`

Comment (OliwierOST) [inline components/EditProfile.tsx:8]: As mentioned before, there should be no default exports except for pages

File: components/EditProfile.tsx [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        ...updatedProfile,
+        id: user.uid,
+      })
+      console.log("success")
+    } catch (error) {
+      alert("Error:" + error)
+    }
+  }
+
+  const initialValues = {
```
Comment (OliwierOST) [inline components/EditProfile.tsx:45]: Instead of `hero ? hero.property` do `hero?.property || ""`

File: components/EditProfile.tsx [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+import { doc, setDoc } from "firebase/firestore"
+import { Caption, FormField } from "./Form"
+
+import { Hero } from "../storage/hero"
+
+export default function EditProfile({ hero }): JSX.Element {
+  const firestore = useFirestore()
+  const { data: user } = useUser()
+
+  const handleSubmit = async (values) => {
```
Comment (OliwierOST) [inline components/EditProfile.tsx:12]: The `values` should have a type defined

File: components/EditProfile.tsx [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    portfolio: hero ? hero.portfolio : "",
+    experience: hero ? hero.experience : [],
+  }
+
+  if (!user) {
+    return <div>loading</div>
+  }
+
+  return (
+    <div>
```
Comment (OliwierOST) [inline components/EditProfile.tsx:66]: We are using MUI component library for the project, please use it here as well

File: components/Profile.tsx [hunk 5]
```diff
@@ -1,24 +1,67 @@
...
+            experience: [
+              { position: "", company: "", startDate: "", endDate: "" },
+            ],
+          })
+      } catch (error) {
+        alert("Error:" + error)
+      }
+    }
+
+    if (status === "success" && user) getHeroData()
```
Comment (OliwierOST) [inline components/Profile.tsx:49]: nit: one-line if statements are not as readable, please split to regular, multi-line

File: components/Profile.tsx [hunk 6]
```diff
@@ -1,24 +1,67 @@
...
-        </div>
-      ) : (
-        <div>err</div>
-      )}
-    </Container>
+    <div>
+      <div>{user?.email}</div>
+      <div onClick={() => auth.signOut()}>logout</div>
+      <div>USER PROFILE</div>
+      <div onClick={() => setEditMode(!editMode)}>
```
Comment (OliwierOST) [inline components/Profile.tsx:61]: As per comment before, this is a redundant component

The user menu is already there in the header

File: components/Profile.tsx [hunk 7]
```diff
@@ -1,24 +1,67 @@
...
+        {!editMode ? "Click to edit Profile" : "cancel edits"}
+      </div>
+      {editMode ? <EditProfile hero={heroData} /> : null}
```
Comment (OliwierOST) [inline components/Profile.tsx:64]: Should be `{editMode && <EditProfile hero={heroData} />}`

File: components/Profile.tsx [hunk 8]
```diff
@@ -1,24 +1,67 @@
...
+import { doc, getDoc } from "firebase/firestore"
+import EditProfile from "./EditProfile"
 
 export default function Profile() {
-  const { data: user } = useUser()
+  const { status, data: user } = useUser()
   const auth = useAuth()
 
+  const [editMode, setEditMode] = useState(false)
+  const [heroData, setHeroData] = useState({})
```
Comment (OliwierOST) [inline components/Profile.tsx:11]: Rename `heroData` to `hero` to stay consistent

The `Hero` component accepts `hero` props, not `heroData`

**Review by OliwierOST — CHANGES_REQUESTED** [review]

File: components/EditProfile.tsx (hunk 1 above)
Comment (rpollock03) [reply components/EditProfile.tsx]: thanks fixed

File: components/EditProfile.tsx (hunk 2 above)
Comment (rpollock03) [reply components/EditProfile.tsx:45]: true, I have changed.

File: components/EditProfile.tsx (hunk 4 above)
Comment (rpollock03) [reply components/EditProfile.tsx]: Cheers added

File: components/Profile.tsx (hunk 5 above)
Comment (rpollock03) [reply components/Profile.tsx:51]: fixed thanks

File: components/Profile.tsx (hunk 6 above)
Comment (rpollock03) [reply components/Profile.tsx]: removed, thanks.

File: components/Profile.tsx (hunk 8 above)
Comment (rpollock03) [reply components/Profile.tsx:13]: done

File: components/EditProfile.tsx (hunk 3 above)
Comment (rpollock03) [reply components/EditProfile.tsx:12]: Don't think we can do that at this stage. Can discuss on slack.

File: components/EditProfile.tsx [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+  return (
+    <Box>
+      <Formik
+        initialValues={initialValues}
+        onSubmit={(values) => handleSubmit(values)}
+      >
+        {({ values, handleChange, handleBlur, handleSubmit, handleReset }) => (
+          <Form onSubmit={handleSubmit}>
+            <Caption>First Name</Caption>
+            <FormField
```
Comment (OliwierOST) [inline components/EditProfile.tsx:74]: As we're using the MUI components library, you should change the FormField and every component possible to the ones coming from MUI. The MUI actually works pretty well with the formik package.

File: components/EditProfile.tsx [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+    return <div>loading</div>
+  }
+
+  return (
+    <Box>
+      <Formik
+        initialValues={initialValues}
+        onSubmit={(values) => handleSubmit(values)}
+      >
+        {({ values, handleChange, handleBlur, handleSubmit, handleReset }) => (
```
Comment (OliwierOST) [inline components/EditProfile.tsx:71]: All of the arguments that the function takes should have a type. The type can be imported directly from formik as FormikProps<ValuesType>

File: components/EditProfile.tsx [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+            />
+            <FieldArray
+              name="experience"
+              render={({ remove, push }) => (
+                <Box>
+                  {typeof values.experience !== "undefined" &&
+                    values.experience.map((item, index) => (
+                      <div key={index}>
+                        <div>
+                          <label htmlFor={`experience.${index}.position`}>
```
Comment (OliwierOST) [inline components/EditProfile.tsx:169]: Instead of label we would rather use the MUI Typography component

Comment (OliwierOST) [inline components/EditProfile.tsx:169]: Instead of label we would rather use the MUI Typography component

File: components/EditProfile.tsx [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+            <FormField
+              type="text"
+              name="website"
+              value={values.website}
+              onChange={handleChange}
+              onBlur={handleBlur}
+            />
+            <FieldArray
+              name="experience"
+              render={({ remove, push }) => (
```
Comment (OliwierOST) [inline components/EditProfile.tsx:163]: Define the type of remove and push, u could either create new interface or maybe import an existing type

File: components/Profile.tsx [hunk 13]
```diff
@@ -1,24 +1,67 @@
...
-          <div>{user.email}</div>
-          <div onClick={() => auth.signOut()}>logout</div>
-        </div>
-      ) : (
-        <div>err</div>
-      )}
-    </Container>
+    <Box>
+      <Button variant="contained" onClick={() => setEditMode(!editMode)}>
+        {" "}
```
Comment (OliwierOST) [inline components/Profile.tsx:60]: is the empty string essential here?

File: components/EditProfile.tsx [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+
+  return (
+    <Box>
+      <Formik
+        initialValues={initialValues}
+        onSubmit={(values) => handleSubmit(values)}
+      >
+        {({ values, handleChange, handleBlur, handleSubmit, handleReset }) => (
+          <Form onSubmit={handleSubmit}>
+            <Caption>First Name</Caption>
```
Comment (OliwierOST) [inline components/EditProfile.tsx:73]: U could wrap the whole column with the MUI Stack component as it works as a flexbox essentially. Try to use it wherever you have evenly spaced components.

File: components/EditProfile.tsx [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+                          <label htmlFor={`experience.${index}.position`}>
+                            Position
+                          </label>
+                          <FormField
+                            name={`experience.${index}.position`}
+                            placeholder="e.g. Senior Developer"
+                            type="text"
+                          />
+                        </div>
+                        <div>
```
Comment (OliwierOST) [inline components/EditProfile.tsx:178]: Same here, every div wrapper could be a Stack component

**Review by OliwierOST — CHANGES_REQUESTED** [review]

**Pull request closed without merging**